Integration tests do not clean up instances under an exception

#### Environment details

spanner/v1.49.2

#### Steps to reproduce

When the following code has a timeout exception, the created instance will not be cleaned up. 

https://github.com/googleapis/java-spanner/blob/ce93f87871e7f42a47a325c1350b4257504c6410/google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestEnv.java#L124

I get the following exception: 

```bash
Caused by: java.util.concurrent.TimeoutException: Waited 30000 milliseconds (plus 129893 nanoseconds delay) for com.google.common.util.concurrent.AbstractCatchingFuture$CatchingFuture@26b95b0b[status=PENDING, info=[inputFuture=[com.google.api.core.ApiFutureToListenableFuture@5f7da3d3], exceptionType=[class java.lang.Exception], fallback=[com.google.api.core.ApiFutures$GaxFunctionToGuavaFunction@103082dd]]]
        at com.google.common.util.concurrent.AbstractFuture.get(AbstractFuture.java:509)
        at com.google.common.util.concurrent.FluentFuture$TrustedFuture.get(FluentFuture.java:102)
        at com.google.common.util.concurrent.ForwardingFuture.get(ForwardingFuture.java:68)
        at com.google.api.gax.longrunning.OperationFutureImpl.get(OperationFutureImpl.java:133)
        at com.google.cloud.spanner.IntegrationTestEnv.initializeInstance(IntegrationTestEnv.java:124)
        ... 21 more
```

But the instance is created successfully but is not deleted. Then, all subsequent tests fail with "Instance already exists" error:

```bash
com.google.cloud.spanner.SpannerException: ALREADY_EXISTS: com.google.cloud.spanner.SpannerException: ALREADY_EXISTS: io.grpc.StatusRuntimeException: ALREADY_EXISTS: Instance already exists: projects/test-project/instances/test-instance
```

#### Expected behaviour

If a test fails with an exception but an instance is created, the instance should be deleted properly and it should not affect the subsequent tests.

--- a/google-cloud-spanner/clirr-ignored-differences.xml
+++ b/google-cloud-spanner/clirr-ignored-differences.xml
@@ -147,4 +147,16 @@
     <method>com.google.api.gax.paging.Page listDatabases()</method>
   </difference>
   
+  <!-- Adding operation RPCs to InstanceAdminClient. -->
+  <difference>
+    <differenceType>7012</differenceType>
+    <className>com/google/cloud/spanner/InstanceAdminClient</className>
+    <method>void cancelOperation(java.lang.String)</method>
+  </difference>
+  <difference>
+    <differenceType>7012</differenceType>
+    <className>com/google/cloud/spanner/InstanceAdminClient</className>
+    <method>com.google.longrunning.Operation getOperation(java.lang.String)</method>
+  </difference>
+  
 </differences>

--- a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceAdminClient.java
+++ b/google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceAdminClient.java
@@ -20,6 +20,7 @@ import com.google.api.gax.longrunning.OperationFuture;
 import com.google.api.gax.paging.Page;
 import com.google.cloud.Policy;
 import com.google.cloud.spanner.Options.ListOption;
+import com.google.longrunning.Operation;
 import com.google.spanner.admin.instance.v1.CreateInstanceMetadata;
 import com.google.spanner.admin.instance.v1.UpdateInstanceMetadata;
 
@@ -217,4 +218,10 @@ public interface InstanceAdminClient {
 
   /** Returns a builder for {@code Instance} object with the given id. */
   Instance.Builder newInstanceBuilder(InstanceId id);
+
+  /** Cancels the specified long-running operation. */
+  void cancelOperation(String name);
+
+  /** Gets the specified long-running operation. */
+  Operation getOperation(String name);
 }

--- a/google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceAdminClientImpl.java
+++ b/google-cloud-spanner/src/main/java/com/google/cloud/spanner/InstanceAdminClientImpl.java
@@ -30,6 +30,7 @@ import com.google.cloud.spanner.SpannerImpl.PageFetcher;
 import com.google.cloud.spanner.spi.v1.SpannerRpc;
 import com.google.cloud.spanner.spi.v1.SpannerRpc.Paginated;
 import com.google.common.base.Preconditions;
+import com.google.longrunning.Operation;
 import com.google.protobuf.FieldMask;
 import com.google.spanner.admin.instance.v1.CreateInstanceMetadata;
 import com.google.spanner.admin.instance.v1.UpdateInstanceMetadata;
@@ -219,5 +220,15 @@ class InstanceAdminClientImpl implements InstanceAdminClient {
   @Override
   public Instance.Builder newInstanceBuilder(InstanceId id) {
     return new Instance.Builder(this, dbClient, id);
+  }
+
+  @Override
+  public void cancelOperation(String name) {
+    rpc.cancelOperation(Preconditions.checkNotNull(name));
+  }
+
+  @Override
+  public Operation getOperation(String name) {
+    return rpc.getOperation(Preconditions.checkNotNull(name));
   }
 }

--- a/google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestEnv.java
+++ b/google-cloud-spanner/src/test/java/com/google/cloud/spanner/IntegrationTestEnv.java
@@ -22,6 +22,8 @@ import com.google.api.gax.longrunning.OperationFuture;
 import com.google.cloud.spanner.testing.RemoteSpannerHelper;
 import com.google.common.collect.Iterators;
 import com.google.spanner.admin.instance.v1.CreateInstanceMetadata;
+import io.grpc.Status;
+import java.util.Random;
 import java.util.concurrent.TimeUnit;
 import java.util.logging.Level;
 import java.util.logging.Logger;
@@ -82,7 +84,10 @@ public class IntegrationTestEnv extends ExternalResource {
       isOwnedInstance = false;
       logger.log(Level.INFO, "Using existing test instance: {0}", instanceId);
     } else {
-      instanceId = InstanceId.of(config.spannerOptions().getProjectId(), "test-instance");
+      instanceId =
+          InstanceId.of(
+              config.spannerOptions().getProjectId(),
+              String.format("test-instance-%08d", new Random().nextInt(100000000)));
       isOwnedInstance = true;
     }
     testHelper = createTestHelper(options, instanceId);
@@ -123,6 +128,34 @@ public class IntegrationTestEnv extends ExternalResource {
     try {
       createdInstance = op.get(30000L, TimeUnit.MILLISECONDS);
     } catch (Exception e) {
+      boolean cancelled = false;
+      try {
+        // Try to cancel the createInstance operation.
+        instanceAdminClient.cancelOperation(op.getName());
+        com.google.longrunning.Operation createOperation =
+            instanceAdminClient.getOperation(op.getName());
+        cancelled =
+            createOperation.hasError()
+                && createOperation.getError().getCode() == Status.CANCELLED.getCode().value();
+        if (cancelled) {
+          logger.info("Cancelled the createInstance operation because the operation failed");
+        } else {
+          logger.info(
+              "Tried to cancel the createInstance operation because the operation failed, but the operation could not be cancelled. Current status: "
+                  + createOperation.getError().getCode());
+        }
+      } catch (Throwable t) {
+        logger.log(Level.WARNING, "Failed to cancel the createInstance operation", t);
+      }
+      if (!cancelled) {
+        try {
+          instanceAdminClient.deleteInstance(instanceId.getInstance());
+          logger.info(
+              "Deleted the test instance because the createInstance operation failed and cancelling the operation did not succeed");
+        } catch (Throwable t) {
+          logger.log(Level.WARNING, "Failed to delete the test instance", t);
+        }
+      }
       throw SpannerExceptionFactory.newSpannerException(e);
     }
     logger.log(Level.INFO, "Created test instance: {0}", createdInstance.getId());

